crates/core/src/factories/core_bootstrap.rs

```diff
@@ -0,0 +1,296 @@
+            let test_server = if server_url.starts_with("test:") {
+                let test_server = TestBootstrapSrv::new(server_url);
+                server_url = test_server.server_address().into();
+                Some(test_server)
+            } else {
+                None
+            };
+
+            (server_url, test_server)
+        };
+
+        let (push_send, push_recv) = tokio::sync::mpsc::channel(1024);
+
+        let push_task = tokio::task::spawn(push_task(
+            config.clone(),
+            server_url.clone(),
+            push_send.clone(),
+            push_recv,
+        ));
+
+        let poll_task = tokio::task::spawn(poll_task(
+            builder,
+            config,
+            server_url,
+            space.clone(),
+            peer_store,
+        ));
+
+        Self {
+            space,
+            push_send,
+            push_task,
+            poll_task,
+            #[cfg(feature = "test_utils")]
+            _test_server,
+        }
+    }
+}
+
+impl Bootstrap for CoreBootstrap {
+    fn put(&self, info: Arc<agent::AgentInfoSigned>) {
+        // ignore puts outside our space.
+        if info.space != self.space {
+            return;
+        }
+
+        // if we can't push onto our large buffer channel... we've got problems
+        let _ = self.push_send.try_send(info);
+    }
+}
+
+async fn push_task(
+    config: CoreBootstrapConfig,
+    server_url: Arc<str>,
+    push_send: PushSend,
+    mut push_recv: PushRecv,
+) {
+    let mut wait = None;
+
+    while let Some(info) = push_recv.recv().await {
+        let url =
+            format!("{server_url}/bootstrap/{}/{}", &info.space, &info.agent);
+        let enc = match info.encode() {
+            Err(_) => continue,
+            Ok(enc) => enc,
+        };
+        match tokio::task::spawn_blocking(move || {
+            ureq::put(&url).send_string(&enc)
+        })
+        .await
+        {
+            Ok(Ok(_)) => {
+                // the put was successful, we don't need to wait
+                // before sending the next info if it is ready
+                wait = None;
+            }
+            _ => {
+                let now = Timestamp::now();
+
+                // the put failed, send it back to try again if not expired
+                if info.expires_at > now {
+                    let _ = push_send.try_send(info);
+                }
+
+                // we need to configure a backoff so we don't hammer the server
+                match wait {
+                    None => wait = Some(config.backoff_min()),
+                    Some(p) => {
+                        let mut p = p * 2;
+                        if p > config.backoff_max() {
+                            p = config.backoff_max();
+                        }
+                        wait = Some(p);
+                    }
+                }
+
+                // wait for the backoff time
+                if let Some(wait) = &wait {
```

Comment:
This is okay because we just have one bootstrapped agent per space? Otherwise we're blocking consuming the queue for the other agent

Reply:
Yep fair enough, that makes sense!